os/src/spsc.rs

```diff
@@ -238,29 +238,12 @@ impl<'q, T> Pusher<'q, T> {
     /// If there is not space, this returns `Err(value)` -- that is, ownership
     /// of `value` is handed back to you.
     pub fn try_push(&mut self, value: T) -> Result<(), T> {
-        let h = self.q.head.load(Ordering::Relaxed);
-        let t = self.q.tail.load(Ordering::Acquire);
-        let h_next = self.q.next_index(h);
-
-        if h_next == t {
-            // We're full.
-            return Err(value);
+        if let Some(entry) = self.try_reserve() {
+            entry.push(value);
+            Ok(())
+        } else {
+            Err(value)
```

Comment:
I felt like I ought to be able to rewrite this to use `Option::ok_or_else`, but it just doesn't work because it'd move `T`. So I think this is probably the most concise way of expressing it already.